spec/components/one_time_code_input_component_spec.rb

```diff
@@ -107,22 +119,50 @@
     end
   end
 
-  describe 'maxlength' do
-    context 'no maxlength given' do
-      it 'renders input maxlength DIRECT_OTP_LENGTH + 1' do
-        expect(rendered).to have_selector(
-          "[maxlength=\"#{TwoFactorAuthenticatable::DIRECT_OTP_LENGTH + 1}\"]",
-        )
+  describe 'code_length' do
+    context 'without code_length' do
+      it 'renders input default maxlength' do
+        expect(rendered).to have_css('[maxlength="6"]')
+      end
+
+      it 'renders input pattern' do
+        expect(rendered).to have_css('[pattern="[0-9]{6}"]')
+      end
+    end
+
+    context 'with code_length' do
+      let(:options) { { code_length: 10 } }
+
+      it 'renders input maxlength based on given code_length' do
+        expect(rendered).to have_css('[maxlength="10"]')
+      end
+
+      it 'renders input pattern' do
+        expect(rendered).to have_css('[pattern="[0-9]{10}"]')
+      end
+    end
+  end
+
+  describe 'optional_prefix' do
+    context 'without optional_prefix' do
+      it 'renders input default maxlength' do
+        expect(rendered).to have_css('[maxlength="6"]')
+      end
+
+      it 'renders input pattern' do
+        expect(rendered).to have_css('[pattern="[0-9]{6}"]')
       end
     end
 
-    context 'maxlength given' do
-      let(:options) { { maxlength: 10 } }
+    context 'with optional_prefix given' do
+      let(:options) { { optional_prefix: '$' } }
+
+      it 'renders input maxlength based on given optional_prefix' do
+        expect(rendered).to have_css('[maxlength="7"]')
+      end
 
-      it 'renders input given maxlength' do
-        expect(rendered).to have_selector(
-          '[maxlength="10"]',
-        )
+      it 'renders input pattern' do
+        expect(rendered).to have_css('[pattern="\\\$?[0-9]{6}"]')
```

Comment:
that is a lot of backslashes!

Reply:
Yeah this might actually be invalid based on the failing spec, gonna take a closer look.

Reply:
That ended up being an interesting deep-dive into differences of regex escaping between Ruby & JavaScript 😅 

It ain't pretty, but 815b1bc1f should address the spec failures.